cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -189,8 +187,9 @@ Status ReplaceWithArrayMask(KernelContext* ctx, const ArrayData& array,
     const ArrayData& array_repl = *replacements.array();
     ReplaceWithArrayMaskImpl<Functor>(
         array, mask, array_repl, replacements_bitmap,
-        CopyArrayBitmap{replacements_bitmap ? array_repl.buffers[0]->data() : nullptr,
-                        array_repl.offset},
+        CopyArrayBitmap{
+            (!!array_repl.buffers[0]) ? array_repl.buffers[0]->data() : nullptr,
```

Comment:
nit: 1) why the double negation? but 2) `replacements_bitmap` (the original condition) is already equivalent or better (MayHaveNulls() will also be false if the null count is 0, which doesn't necessarily imply that the validity bitmap is null).